### ISSUE TITLE
1.1 Todo list

### ISSUE THREAD

**Issue by denizyuret** [denizyuret/Knet.jl#337]
- [ ] conv.jl memory management and algorithm selection issues.
- [x] figure out why p.ptr=C_NULL does not work in freeKnetPtr. test with fill(0) during free.
- [x] pull loss.jl
- [ ] test ekin's slowdown example (array overwrite).
- [x] fix serialization.
- [x] replace constants with environment variables so they can be configured without code change.
- [ ] pull requests.
- [x] Knet profiling based on TimerOutputs in macros similar to AutoGrad.
- [x] Document AutoGrad profiling features.
- [ ] kptr check ages in mac model to see if free would be useful.
- [ ] check cuda 10 and cudnn 7.3  for performance improvements.
- [x] Train interface only does nll, is not general enough.
- [ ] AutoGrad only uses broadcasted now, compare performance with using broadcast
- [ ] add using LinearAlgebra: lmul!, rmul! to test/linalg.jl
- [ ] test on other AD and GPUarray pkgs.
- [x] Add example models / layers for Linear, MLP, CNN, RNN to Knet ???
- [ ] Update docs and tests with new model interface and training related functions, check test coverage.
- [ ] separate and move KnetArray to KnetML.
- [ ] try making karray <: AbstractArray and overriding get/setindex.
- [ ] eventually, slim down update! and rnn gpu tests
- [x] limit max memory allocated by kptr
- [ ] unsafe_copy! is not in base any more? fix unsafe_copy!, unsafe_convert in karray looking in base
- [ ] unary_nd, indexed_function, isequivalent, _dbg, ssize not in AutoGrad any more.
- [ ] reorganize unary.jl and broadcast.jl
- [ ] check TODO.
- [ ] reconsider Requires.jl.
- [ ] Knet/tutorial
  - [x] 00.Julia_is_fast.ipynb
  - [x] 01.Getting_to_know_Julia.ipynb
  - [x] 02.mnist.ipynb
  - [ ] 03.lin.ipynb: use this as a mini AutoGrad demo? start with functional style? (https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/knet-users/dwK1bPHV36o/nTLJoqpRCwAJ)
  - [x] 04.mlp.ipynb
  - [ ] 05.cnn.ipynb: let user draw own digits.
  - [ ] 06.rnn.ipynb: ~~add figure~~, manual rnn definition, keepstate reduces performance?
  - [x] 07.imdb.ipynb: let user ask own questions.
  - [ ] 08.charlm.ipynb: add the julia model.
  - [ ] 09.s2s.ipynb: convert to new interface, do real MT?
  - [ ] check prerequisites between notebooks
  - [ ] add optimization tutorial
  - [x] place tutorial in a more prominent location
  - [x] update README.md
  - [x] figure out how to save with plots.
  - [x] add option to download models.
  - [x] make all work without gpu
  - [x] add ProgressMeter to auto-install list.
- [ ] Knet/docs
  - [ ] backprop.md
  - [ ] cnn.md
  - [ ] gen.md
  - [ ] index.md
  - [ ] install.md
  - [ ] mlp.md
  - [ ] opt.md
  - [ ] reference.md
  - [ ] rl.md
  - [ ] rnn.md
  - [ ] softmax.md
  - [ ] tutorial.md
- [ ] Knet/examples
  - [ ] update all to use the new interface.
  - [X] charlm: deprecate, in tutorial
  - [X] cifar10-cnn
  - [X] dcgan-mnist
  - [ ] DeepLearningFrameworks: submit new results.
  - [X] dl-tutorial: moved to Knet/tutorial
  - [x] dynet-benchmark
  - [X] fashion-mnist
  - [X] housing-linreg
  - [ ] julia-tutorial: check old commands, turn into notebook.
  - [x] knet-tutorial: deprecate -> Knet/tutorial
  - [X] lenet
  - [X] mnist-mlp
  - [X] optimizers
  - [x] overfitting: deprecate this, it is replicated in tutorial.
  - [x] reinforcement-learning
  - [X] resnet
  - [x] rnnlm: update this with new interface, check for performance regression
  - [x] rnn-tutorial: check for performance regression. ~~deprecate after Knet/tutorial completed.~~ keep it as handwritten lstm example, replace tutorial notebook instead.
  - [X] synthetic-linreg
  - [X] variational-autoencoder
  - [X] vgg
- [ ] KnetML: check projects for Julia 1.0 compatibility
  - [ ] MAC-Network
  - [ ] NLPDemos
  - [ ] WGAN
  - [ ] Relational Network
  - [ ] Knet-the-Julia-dope
  - [ ] singularity-images
  - [ ] Neural-Style-Transfer
- [ ] Knet/prof
  - [x] broadcast.jl
  - [ ] conv.jl: check regression in poolx(::Array), conv4x(::Array)
  - [ ] karray.jl: check regression in vcat(mat,row), vcat(row,mat), hcat(kcol,kmat), bad 0.6 tests for vcat(kmat,krow)
  - [x] profflat.pl
  - [x] proftree.pl
  - [ ] reduction.jl: check regression in sum(a,0), slight regression in sum(a,1) and sum(a,2)
  - [x] rnnlm.jl: gc regression in for loop tests
  - [x] s2s.jl: **100x regression** 
  - [x] softmax.jl 
- [x] Add links for KnetML, knet-dev, knet-users etc in README.md.
- [x] use global keyword in the for loops in tests
- [x] update travis.yml (and even better add gpu testing through #312)
- [x] add Project.toml
- [x] add Manifest.toml to .gitignore
- [x] update readme badges
- [x] data
- [x] deps
- [x] src: CPU:20/20
- [x] test: CPU:14/14
- [x] fix JLD.
- [x] fix KnetDisplay (summary line should show KnetArray) and other display/show problems.
- [x] seed! gc dir: just use the same names but have Knet versions.
- [x] new AutoGrad interface.
